src/common/include/sirf/common/jcg.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <cmath>
+#include <iostream>
+
+#include <boost/shared_ptr.hpp>
+
+template<class vector>
+class anOperator {
+public:
+	virtual std::shared_ptr<vector> apply(const vector& v) const = 0;
+	std::shared_ptr<vector> operator()(const vector& v) const
+	{
+		return apply(v);
+	}
+};
+
+template<class value>
+class JacobiCG {
+public:
+	JacobiCG() : nit_(10) {}
+	void set_num_iterations(int nit)
+	{
+		nit_ = nit;
+	}
+	template<class vector>
+	value rightmost(const anOperator<vector>& A, vector& x) const
+	{
+		value lmd;
+		value a[] = { 0, 0, 0, 0 };
+		value mu[2];
+		value u[2];
+		value v[2];
+
+		std::unique_ptr<vector> uptr_y = x.clone();
+		vector& y = *uptr_y;
+		std::unique_ptr<vector> uptr_z = x.clone();
+		vector& z = *uptr_z;
+		std::unique_ptr<vector> uptr_w = x.clone();
+		vector& w = *uptr_w;
+		std::unique_ptr<vector> uptr_Az = x.clone();
+		vector& Az = *uptr_Az;
+
+		value s = sqrt(x.dot(x));
+		x.scale(s);
+		std::shared_ptr<vector> sptr_Ax = A(x);
+		vector& Ax = *sptr_Ax;
+
+		for (int it = 0; it < nit_; it++) {
+			lmd = Ax.dot(x);
+			y.axpby(1.0, Ax, -lmd, x);
+			std::cout << it << ": " << lmd << '\n';
```

Comment:
best to remove this later, or at least say more

Reply:
commented out for now, ideally `verbosity` needed